**Laboratorio 6**

María Paula Angulo Collantes

Danna Gabriela Castro

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
df = pd.read_csv('bicicletas_compartidas_bogota.csv')

**Punto 1**

In [2]:
sub = df.dropna(subset=['distancia_km','duracion_min'])
x = sub['distancia_km'].to_numpy()
y = sub['duracion_min'].to_numpy()

cov = np.cov(x, y, ddof=1)[0, 1]

r = cov / (x.std(ddof=1) * y.std(ddof=1))
print(f"Pearson r: {r:.4f}")


Pearson r: 0.7217


**1.b** 

**Dirección:** positiva, a mayor distancia recorrida, tiende a haber mayor duración del viaje.

**Intensidad:** moderada-alta. Un r de 0.72 no es una correlación perfecta, pero sí es considerablemente fuerte para datos de comportamiento humano/urbano, donde hay mucha variabilidad.

**Significado práctico:** R²≈0.52. El 52% de la variación en duración se explica linealmente por la distancia recorrida. El 48% restante depende de otros factores no capturados aquí.

**1.c**

No es correcto afirmarlo únicamente con base en la correlación, porque esta muestra asociación, no causalidad. Aunque existe una relación física razonable entre distancia y duración, podrían influir otras variables como el clima o el tipo de bicicleta. Por eso, lo adecuado es decir que existe una asociación positiva entre ambas variables, no que la distancia causa directamente la duración.

**1.d**

No. Un coeficiente de correlación como r=0.72 resume una tendencia general calculada sobre miles de observaciones, pero no garantiza el comportamiento de cada viaje individual. Es un promedio estadístico, no una regla determinística.

**Justificación:** estos casos muestran que factores no capturados en la correlación pueden hacer que un viaje individual se aleje completamente del patrón general. La correlación describe cómo se comporta el conjunto de datos en promedio, pero viajes puntuales pueden romper esa tendencia sin que eso invalide el hallazgo global.

**Punto 2**

In [3]:
q1_dist=df['distancia_km'].quantile(0.25)
q1_dur=df['duracion_min'].quantile(0.25)
print(q1_dist)
print(q1_dur)


1.71
8.9


In [4]:
q3_dist=df['distancia_km'].quantile(0.75)
q3_dur=df['duracion_min'].quantile(0.75)
print(q3_dist)
print(q3_dur)


3.58
17.6


In [5]:
iqr_dist=q3_dist-q1_dist
iqr_dur=q3_dur-q1_dur
print(iqr_dist)
print(iqr_dur)

1.87
8.700000000000001


In [6]:
lim_inferior_dist = q1_dist - 1.5 * iqr_dist
lim_inferior_dur = q1_dur - 1.5 * iqr_dur
print(lim_inferior_dist)
print(lim_inferior_dur)

-1.0950000000000002
-4.15


In [7]:
lim_superior_dist = q3_dist + 1.5 * iqr_dist
lim_superior_dur = q3_dur + 1.5 * iqr_dur
print(lim_superior_dist)
print(lim_superior_dur)

6.385
30.650000000000002


In [8]:
outliers_dist = sub[(sub['distancia_km'] < lim_inferior_dist) | (sub['distancia_km'] > lim_superior_dist)]
outliers_dur  = sub[(sub['duracion_min'] < lim_inferior_dur) | (sub['duracion_min'] > lim_superior_dur)]
print(outliers_dist)
print(outliers_dur)

     viaje_id         inicio_viaje        estacion_origen localidad_origen  \
10    V001245  2026-03-03 07:34:14   Parque de los Novios   Barrios Unidos   
16    V000965  2026-04-10 17:36:36  Parque de los Hippies        Chapinero   
35    V001542  2026-02-17 08:25:32   Universidades Centro    La Candelaria   
70    V000670  2026-03-01 18:51:29               Calle 72        Chapinero   
113   V002582  2026-05-16 08:50:59   Universidades Centro    La Candelaria   
...       ...                  ...                    ...              ...   
2993  V002477  2026-04-19 20:32:38               Calle 72        Chapinero   
3031  V001949  2026-04-07 18:23:01    Simón Bolívar Norte   Barrios Unidos   
3079  V000727  2026-04-05 19:01:40              Calle 100          Usaquén   
3133  V001108  2026-03-09 07:10:58  Universidad Javeriana        Chapinero   
3203  V000346  2026-02-21 23:08:18              Corferias      Teusaquillo   

       estacion_destino localidad_destino  duracion_min  distan

In [9]:
mask_outliers = ((sub['distancia_km'] < lim_inferior_dist) | (sub['distancia_km'] > lim_superior_dist) |
                  (sub['duracion_min'] < lim_inferior_dur) | (sub['duracion_min'] > lim_superior_dur))

outliers_totales = sub[mask_outliers]
print(f"Total de registros atípicos (en al menos una variable): {len(outliers_totales)} de {len(sub)} ({100*len(outliers_totales)/len(sub):.2f}%)")

Total de registros atípicos (en al menos una variable): 147 de 3162 (4.65%)


In [18]:
sub = df.dropna(subset=['distancia_km','duracion_min']).copy()
sub

,viaje_id,inicio_viaje,estacion_origen,localidad_origen,estacion_destino,localidad_destino,duracion_min,distancia_km,tipo_bicicleta,tipo_usuario,clima,temperatura_c,precipitacion_mm,fin_de_semana
0,V000671,2026-04-22 08:35:33,Plaza de Bolívar,La Candelaria,Universidades Centro,La Candelaria,13.8,4.14,Eléctrica,Suscriptor,Nublado,17.2,0.0,No
1,V002336,2026-03-26 20:39:41,Plaza de Bolívar,La Candelaria,Universidades Centro,La Candelaria,10.8,2.47,Eléctrica,Ocasional,Lluvia ligera,15.2,0.9,No
2,V002504,2026-03-09 11:25:42,Parque de los Novios,Barrios Unidos,Hacienda Santa Bárbara,Usaquén,6.6,0.83,Eléctrica,Suscriptor,Nublado,17.1,0.0,No
3,V002587,2026-04-29 04:04:46,Parque de la 93,Chapinero,Parque de la 93,Chapinero,7.6,1.68,Mecánica,Suscriptor,Nublado,21.0,0.0,No
4,V000820,2026-05-26 03:55:26,Zona G,Chapinero,Parque Nacional,Santa Fe,16.6,2.89,Mecánica,Suscriptor,Lluvia,18.7,7.8,No
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3209,V001604,2026-02-09 08:38:47,Centro Internacional,Santa Fe,Universidad Javeriana,Chapinero,13.9,3.06,Eléctrica,Suscriptor,Nublado,15.1,0.0,No
3210,V002551,2026-03-21 15:54:26,Parque de los Novios,Barrios Unidos,Parque de los Novios,Barrios Unidos,6.0,1.49,electrica,Ocasional,Nublado,16.9,0.0,Sí
3211,V000538,2026-04-20 20:31:42,San Felipe,Barrios Unidos,Simón Bolívar Norte,Barrios Unidos,16.4,2.23,Eléctrica,Ocasional,Despejado,11.2,0.0,No
3212,V001221,2026-05-15 18:30:35,Parkway,Teusaquillo,Calle 100,Usaquén,8.3,0.97,Mecánica,Suscriptor,Lluvia ligera,18.4,0.3,No


In [21]:
sub = sub.copy()

sub['velocidad_kmh'] = sub['distancia_km'] / (sub['duracion_min'] / 60)
sub['exceso_dist'] = (sub['distancia_km'] - lim_superior_dist).clip(lower=0) / iqr_dist
sub['exceso_dur']  = (sub['duracion_min'] - lim_superior_dur).clip(lower=0) / iqr_dur
sub['severidad'] = sub['exceso_dist'] + sub['exceso_dur']

top5 = sub.sort_values('severidad', ascending=False).head(5)
top5[['distancia_km', 'duracion_min', 'velocidad_kmh']]

,distancia_km,duracion_min,velocidad_kmh
750,3.62,180.0,1.206667
2972,3.71,180.0,1.236667
2089,31.20,8.9,210.337079
1052,27.50,22.3,73.991031
1567,4.65,125.0,2.232000


In [12]:
ids_top5 = top5.index
detalle = df.loc[ids_top5, ['viaje_id', 'inicio_viaje', 'distancia_km', 'duracion_min',
                             'tipo_bicicleta', 'tipo_usuario', 'clima', 'precipitacion_mm']]
detalle['velocidad_kmh'] = detalle['distancia_km'] / (detalle['duracion_min'] / 60)
detalle

,viaje_id,inicio_viaje,distancia_km,duracion_min,tipo_bicicleta,tipo_usuario,clima,precipitacion_mm,velocidad_kmh
750,V002333,2026-04-05 19:05:51,3.62,180.0,Mecánica,Suscriptor,Nublado,0.0,1.206667
2972,V000070,2026-03-28 09:21:27,3.71,180.0,Mecánica,Suscriptor,Despejado,0.0,1.236667
2089,V000844,2026-03-03 20:53:19,31.20,8.9,Mecánica,Suscriptor,Despejado,0.0,210.337079
1052,V002209,2026-05-11 09:21:29,27.50,22.3,Eléctrica,Suscriptor,Lluvia,11.6,73.991031
1567,V001789,2026-05-25 10:27:40,4.65,125.0,Eléctrica,Suscriptor,Despejado,0.0,2.232000


**Punto 3**

In [13]:
sub_sin_outliers = sub[~mask_outliers].copy()

r_con = np.corrcoef(sub['distancia_km'], sub['duracion_min'])[0, 1]
r_sin = np.corrcoef(sub_sin_outliers['distancia_km'], sub_sin_outliers['duracion_min'])[0, 1]

print(f"Pearson r CON outliers (n={len(sub)}): {r_con:.4f}")
print(f"Pearson r SIN outliers (n={len(sub_sin_outliers)}): {r_sin:.4f}")

Pearson r CON outliers (n=3162): 0.7217
Pearson r SIN outliers (n=3015): 0.8887


**3.a**

Sí. La correlación subió de 0.72 a 0.89, un aumento de 0.167 puntos (~23% de cambio relativo). El R² casi se duplica en términos de varianza explicada (de 52% a 79%). Para un cambio producido por solo el 4.65% de las observaciones, es un movimiento importante.

**3.b**

La relación entre distancia y duración parece ser genuina y no un resultado causado únicamente por los valores extremos, ya que la correlación mantiene su dirección positiva y, al eliminar los outliers, incluso se fortalece. Esto indica que el patrón observado está presente en el comportamiento general de los usuarios y no depende solamente de unos pocos registros atípicos.

Sin embargo, el cambio en el coeficiente muestra que la intensidad de la relación es sensible al tratamiento de los datos. Por lo tanto, aunque la asociación es robusta, el valor exacto de Pearson no debe considerarse completamente estable, pues depende de cómo se manejen los registros extremos y de las decisiones de limpieza del dataset.

**3.c**

No. Lo más honesto es reportar ambas, con su respectivo contexto:

Reportar solo r=0.72 (con outliers) sin decir nada más deja pasar valores de velocidad físicamente imposibles (ej. 210 km/h) como si fueran datos válidos, subestimando artificialmente la fuerza real de la relación entre las dos variables.
Reportar solo r=0.89 (sin outliers) sin explicar el criterio de exclusión oculta que se descartó casi el 5% de los datos, y un lector no podría evaluar si esa decisión fue razonable o si se "maquilló" el resultado para que se viera más bonito.

Lo correcto es presentar ambos números junto con la justificación de por qué difieren (identificación de errores de captura probables como el caso de 31.2 km en 8.9 min, frente a casos ambiguos o probablemente válidos como las pausas de 180 min). Esto le da al lector la información completa para juzgar la robustez del hallazgo, en vez de una cifra aislada que oculta la decisión metodológica detrás de ella.

**Punto 4**

In [14]:
bici = df.groupby('tipo_bicicleta').agg(
    n_viajes=('viaje_id', 'count'),
    distancia_prom_km=('distancia_km', 'mean'),
    duracion_prom_min=('duracion_min', 'mean'),
    duracion_mediana_min=('duracion_min', 'median')
).round(2)
bici

,n_viajes,distancia_prom_km,duracion_prom_min,duracion_mediana_min
tipo_bicicleta,,,,
Eléctrica,1511,2.81,12.90,11.40
Mecánica,1678,2.89,15.38,13.55
electrica,25,3.21,17.19,12.50


**Punto 5**

In [15]:
tablaclimas = df.groupby('clima').agg(
    n_viajes=('viaje_id', 'count'),
    distancia_prom_km=('distancia_km', 'mean'),
    duracion_prom_min=('duracion_min', 'mean'),
    duracion_mediana_min=('duracion_min', 'median')
).round(2)
print(tablaclimas)

               n_viajes  distancia_prom_km  duracion_prom_min  \
clima                                                           
Despejado          1337               2.85              14.27   
Lluvia              358               2.94              14.31   
Lluvia ligera       444               2.90              14.56   
Nublado            1059               2.80              14.04   
lluvia ligera        16               2.63              12.67   

               duracion_mediana_min  
clima                                
Despejado                     12.50  
Lluvia                        12.75  
Lluvia ligera                 12.90  
Nublado                       12.30  
lluvia ligera                 12.95  


In [16]:
filtrado = df[(df['distancia_km'] >= 2) & (df['distancia_km'] <= 4)]
tabladosycuatroclima = filtrado.groupby('clima').agg(
    n_viajes=('viaje_id', 'count'),
    distancia_prom_km=('distancia_km', 'mean'),
    duracion_prom_min=('duracion_min', 'mean'),
    duracion_mediana_min=('duracion_min', 'median')
).round(2)
print(tabladosycuatroclima)

               n_viajes  distancia_prom_km  duracion_prom_min  \
clima                                                           
Despejado           607               2.83              14.51   
Lluvia              162               2.87              14.47   
Lluvia ligera       209               2.88              14.44   
Nublado             490               2.81              14.52   
lluvia ligera         3               3.18              14.27   

               duracion_mediana_min  
clima                                
Despejado                     13.90  
Lluvia                        13.85  
Lluvia ligera                 13.70  
Nublado                       13.65  
lluvia ligera                 14.20  


¿Cambian las diferencias entre condiciones climáticas?
Si, cambian según cómo se mire. Sin controlar por distancia, la duración promedio varía muy poco entre climas (apenas 0.5 min de diferencia). Pero al controlar la distancia comparando solo viajes de 2 a 4 km, las diferencias se reducen todavía más y ya no muestran ningún patrón claro.

¿Por qué comparar todos los viajes directamente podría resultar engañoso?
Porque no todos los climas tienen la misma mezcla de distancias por decirlo asi. Si en un clima hay más viajes largos que en otro, ese clima va a mostrar duración promedio más alta, pero no por el clima en si, sino por la distancia. Comparar en bruto relaciones distintas y hacer parecer que el clima influye en esto no es muy adecuado, ya que en realidad es la distancia la que está jalando el resultado.


¿Qué papel puede estar desempeñando distancia km?
El de variable de confusión. Es la que realmente explica la mayor parte de la variación en duración (r= 0.72–0.89), y como probablemente varía también según el clima, genera una relación aparente entre clima y duración que en realidad no es directa.


¿Podemos concluir que el clima causa cambios en la duración del viaje?
No,al controlar por distancia, la asociación entre clima y duración se debilita hasta casi desaparecer. Con estos datos solo se puede hablar de una asociación muy débil, no de una relación causal. Eso sí esto solo aplica a la duración, lo único es que queda pendiente ver si el clima afecta la cantidad de viajes que la gente hace.

**Punto 6**

Lo que más explica cuánto dura un viaje es la distancia: entre más kilómetros, más minutos (r = 0.72 con todos los datos, y sube a r = 0.89 si se quitan los casos raros). El clima, en cambio, casi no influye en la duración una vez que se compara viajes de distancia similar, probablemente parece tener efecto solo porque va "de la mano" con la distancia, no porque realmente la cause. Ademas, hay un dato sospechoso, dos viajes duran exactamente 180.0 minutos. Que coincidan tan exacto sugiere que el sistema les puso un tope automático, tal vez la bici no se devolvió a tiempo y no estrictamente esos 180 min hayan sido el tiempo real de uso, esto habría que confirmarlo con el operador o el usuario. Por otro lado, no podemos decir que "el clima no afecta nada" solo se vio que no afecta cuánto dura un viaje. Falta revisar si afecta cuántos viajes se hacen (la demanda). Ya como recomendaciones, estarías: normalizar los textos inconsistentes (mayúsculas, tildes), revisar y corregir velocidades imposibles y confirmar con el operador si existe un tope de 180 minutos.